### PR TITLE
feat: fetch usernames for waiting page

### LINKED ISSUE #88: Fetch usernames from userId for waiting page

### PR DESCRIPTION
**Pull request by ypinhsuan**
Closes #88

### CODE REVIEW

File: frontend/src/hooks/useUsers.ts [hunk 1]
```diff
@@ -15,3 +16,14 @@ export const useCreateUser = () => {
...
+export const useGetUsername = () => {
+  const getUsername = useCallback(async (userId: number) => {
+    const { data } = await ApiService.get<GetUserNameRes | undefined>(
+      `/users/${userId}`,
+    );
+    return data?.name;
+  }, []);
+
+  return { getUsername };
+};
```
Comment (dextertanyj) [inline frontend/src/hooks/useUsers.ts:29]: Let's use `react-query` to help us cache this:
```suggestion
export const useGetUserName = (userId?: number) => {
  const getUserName = async () => {
    const { data } = await ApiService.get<GetUserNameRes | undefined>(
      `/users/${userId}`,
    );
    return data;
  };
  const { data: user, isLoading: isGetUserNameLoading } = useQuery(
    ["USER", userId?.toString() || ""],
    getUserName,
    {
      enabled: !!userId,
    },
  );
  return {
    user,
    isGetUserNameLoading,
  };
};
```

File: frontend/src/pages/WaitingPage.tsx [hunk 2]
```diff
@@ -4,12 +4,15 @@ import { CircularProgress, Typography } from "@mui/material";
...
 
 import { useSocket } from "src/contexts/WsContext";
+import { useGetUsername } from "src/hooks/useUsers";
 
 import { Match } from "~shared/types/api/match.dto";
 
 export const WaitingPage = () => {
   const navigate = useNavigate();
   const { socket, connect } = useSocket();
+  const { getUsername } = useGetUsername();
```
Comment (dextertanyj) [inline frontend/src/pages/WaitingPage.tsx:14]: If we use `react-query`, then we will need to `useState` to fire off the queries when the user IDs become available.

```suggestion
  const [userOne, setUserOne] = useState<number | undefined>(undefined);
  const [userTwo, setUserTwo] = useState<number | undefined>(undefined);
  const [roomId, setRoomId] = useState<string | undefined>(undefined);
  const { user: userNameOne } = useGetUserName(userOne);
  const { user: userNameTwo } = useGetUserName(userTwo);
```

File: frontend/src/pages/WaitingPage.tsx [hunk 3]
```diff
@@ -47,9 +50,11 @@ export const WaitingPage = () => {
       setTimeout(() => navigate("/dashboard"), 3000);
     }, 30000);
 
-    socket.on("found", (match: Match) => {
+    socket.on("found", async (match: Match) => {
```
Comment (dextertanyj) [inline frontend/src/pages/WaitingPage.tsx:53]: Then we can use this to set the states when we receive the event:
```
      setUserOne(match.result[0].userId);
      setUserTwo(match.result[1].userId);
      setRoomId(match.roomId);
      setMessage("Match found. Loading information...");
      clearTimeout(timeout);
```

File: frontend/src/pages/WaitingPage.tsx [hunk 4]
```diff
@@ -4,12 +4,15 @@ import { CircularProgress, Typography } from "@mui/material";
...
 import { useSocket } from "src/contexts/WsContext";
+import { useGetUsername } from "src/hooks/useUsers";
 
 import { Match } from "~shared/types/api/match.dto";
 
 export const WaitingPage = () => {
   const navigate = useNavigate();
   const { socket, connect } = useSocket();
+  const { getUsername } = useGetUsername();
+
```
Comment (dextertanyj) [inline frontend/src/pages/WaitingPage.tsx:15]: And we can use a `useEffect` hook to set the message when all the data is available.
```
  useEffect(() => {
    if (roomId && userNameOne && userNameTwo) {
      setMessage(
        `Found a match between ${userNameOne.name} and ${userNameTwo.name}!
        Room ID: ${roomId}
        Loading...`,
      );
    }
  }, [roomId, userNameOne, userNameTwo]);
```

**Review by dextertanyj — CHANGES_REQUESTED** [review]
Some comments about how we can use `react-query`.

**Comment by ypinhsuan** [comment]
Thanks for the guide! I was quite confused about how to use react hook inside a callback 😅

**Review by dextertanyj — APPROVED** [review]
lgtm

**Pull request merged**